File: platforms/registry.py
"""The platform registry: the single source of truth for the seven platforms the LabeloxAV data engine hosts.

LabeloxAV is the host. The annotation core and the six folded subsystems (SANYX, CALYX, SIEVYX, ORACLYX,
VERDYX, FORGYX) are Platforms: distinct navigable UIs behind a launcher, over one shared backend spine.

This registry is mirrored on the frontend (web/platforms/registry.ts). Both must be kept in sync; the API
exposes this one at GET /api/platforms so the two can be reconciled in a test. Nothing about the existing
routers moves; a platform simply groups the domain routers that already exist (backend_prefix is the new
canonical mount, legacy_prefixes are the existing ones that keep working). ``gate`` marks a platform that can
block a session's progression through the flywheel (SANYX health, CALYX calibration) or a model's promotion
(VERDYX eval, FORGYX benchmark).
"""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass(frozen=True)
class Platform:
    id: str
    label: str
    role: str                                   # one-line description shown in the launcher
    backend_prefix: str                         # new canonical API prefix, e.g. /sanyx
    legacy_prefixes: tuple[str, ...] = ()        # existing domain routers this platform groups (kept working)
    gate: str | None = None                     # health | calibration | eval | benchmark | None
    order: int = 0                              # left-rail / launcher ordering
    flywheel_stage: int | None = None           # position in the closed loop (section 4), None = not a stage

    def as_dict(self) -> dict:
        return {
            "id": self.id, "label": self.label, "role": self.role,
            "backend_prefix": self.backend_prefix, "legacy_prefixes": list(self.legacy_prefixes),
            "gate": self.gate, "order": self.order, "flywheel_stage": self.flywheel_stage,
        }


PLATFORMS: tuple[Platform, ...] = (
    Platform(
        id="labelox", label="Labelox", role="annotation core: three-path auto-label + human workspace",
        backend_prefix="/labelox", order=0, flywheel_stage=4,
        legacy_prefixes=("/autolabel", "/objects", "/objects3d", "/review", "/corrections", "/tracks",
                         "/multicam", "/lanes", "/segmentation", "/ocr", "/signs", "/triage", "/recall"),
    ),
    Platform(
        id="sanyx", label="SANYX", role="ingest QA: health score, quarantine bad sessions",
        backend_prefix="/sanyx", legacy_prefixes=("/inspector", "/adverse"), gate="health",
        order=1, flywheel_stage=1,
    ),
    Platform(
        id="calyx", label="CALYX", role="calibration monitor: extrinsic / intrinsic drift",
        backend_prefix="/calyx", legacy_prefixes=("/calibration",), gate="calibration",
        order=2, flywheel_stage=2,
    ),
    Platform(
        id="sievyx", label="SIEVYX", role="curation and mining: embed, rank, dedup, decide what to label",
        backend_prefix="/sievyx", order=3, flywheel_stage=3,
        legacy_prefixes=("/curation", "/search", "/intelligence", "/activelearn", "/discovery"),
    ),
    Platform(
        id="oraclyx", label="ORACLYX", role="offline-fusion pseudo-GT: auto-truth the majority, route disagreements",
        backend_prefix="/oraclyx", legacy_prefixes=("/lidar", "/lidarscene", "/hdmap", "/mapassist"),
        order=4, flywheel_stage=5,
    ),
    Platform(
        id="verdyx", label="VERDYX", role="slice evaluation: per-slice regression + champion/challenger verdict",
        backend_prefix="/verdyx", legacy_prefixes=("/training", "/govern", "/models"), gate="eval",
        order=5, flywheel_stage=7,
    ),
    Platform(
        id="forgyx", label="FORGYX", role="edge optimization: quantize, compile, benchmark, gate on latency and accuracy",
        backend_prefix="/forgyx", legacy_prefixes=("/export",), gate="benchmark",
        order=6, flywheel_stage=8,
    ),
)

_BY_ID = {p.id: p for p in PLATFORMS}


def platform_by_id(pid: str) -> Platform | None:
    return _BY_ID.get(pid)


def platform_for_route(path: str) -> Platform | None:
    """Resolve the platform that owns an API path, by its canonical or a legacy prefix."""
    for p in PLATFORMS:
        if path.startswith(p.backend_prefix) or any(path.startswith(lp) for lp in p.legacy_prefixes):
            return p
    return None


def gates() -> list[Platform]:
    """Platforms that can block progression, in flywheel order (the DAG consults these)."""
    return sorted((p for p in PLATFORMS if p.gate), key=lambda p: p.order)


def as_dicts() -> list[dict]:
    return [p.as_dict() for p in sorted(PLATFORMS, key=lambda p: p.order)]

File: web/platforms/registry.ts
// The platform registry (frontend mirror of platforms/registry.py). LabeloxAV is the host; the annotation
// core and the six folded subsystems are Platforms: distinct navigable UIs behind a launcher, over one shared
// backend spine. This drives the launcher, the platform switcher, and the cross-platform session view.
//
// Operational Materialism: a platform tile is neutral graphite. Color is earned only by live state (a SANYX
// quarantine count, a CALYX block, a VERDYX pending verdict), never by decoration. Keep this file in sync with
// platforms/registry.py; GET /api/platforms exposes the backend copy so the two can be reconciled.

export type PlatformId = "labelox" | "sanyx" | "calyx" | "sievyx" | "oraclyx" | "verdyx" | "forgyx";
export type PlatformGate = "health" | "calibration" | "eval" | "benchmark" | null;
export type PlatformNavItem = { href: string; label: string; hint?: string };

export type Platform = {
  id: PlatformId;
  label: string;
  role: string;
  glyph: string;            // short mono rail glyph
  order: number;
  gate: PlatformGate;       // this platform can block progression / promotion
  flywheelStage: number | null;
  home: string;             // route the launcher tile opens
  nav: PlatformNavItem[];   // destinations within the platform (existing routes today)
};

export const PLATFORMS: Platform[] = [
  {
    id: "labelox", label: "Labelox", role: "annotation core: three-path auto-label + human workspace",
    glyph: "LBX", order: 0, gate: null, flywheelStage: 4, home: "/",
    nav: [
      { href: "/", label: "Triage", hint: "object queue ranked by value" },
      { href: "/agent", label: "Agent Console", hint: "autonomous QA and fix queue" },
      { href: "/review/queue", label: "Review queue", hint: "active learning + error candidates" },
      { href: "/annotations", label: "Annotations", hint: "browse and resume sessions" },
      { href: "/jobs", label: "Jobs", hint: "import, training, autolabel runs" },
    ],
  },
  {
    id: "sanyx", label: "SANYX", role: "ingest QA: health score, quarantine bad sessions",
    glyph: "SNX", order: 1, gate: "health", flywheelStage: 1, home: "/inspect",
    nav: [
      { href: "/inspect", label: "Inspector", hint: "MCAP session inspection" },
      { href: "/sanyx", label: "Ingest board", hint: "health scores and quarantine (M1)" },
    ],
  },
  {
    id: "calyx", label: "CALYX", role: "calibration monitor: extrinsic / intrinsic drift",
    glyph: "CLX", order: 2, gate: "calibration", flywheelStage: 2, home: "/calibration",
    nav: [{ href: "/calibration", label: "Calibration", hint: "camera validation and drift" }],
  },
  {
    id: "sievyx", label: "SIEVYX", role: "curation and mining: embed, rank, dedup, decide what to label",
    glyph: "SVX", order: 3, gate: null, flywheelStage: 3, home: "/curation",
    nav: [
      { href: "/curation", label: "Curation", hint: "frame-level active learning" },
      { href: "/search", label: "Search", hint: "visual and semantic similarity" },
      { href: "/discovery", label: "Discovery", hint: "rare-scenario novelty queue" },
      { href: "/scenarios", label: "Scenarios", hint: "behavioral scenario mining" },
    ],
  },
  {
    id: "oraclyx", label: "ORACLYX", role: "offline-fusion pseudo-GT: auto-truth the majority, route disagreements",
    glyph: "ORX", order: 4, gate: null, flywheelStage: 5, home: "/lidar",
    nav: [
      { href: "/lidar", label: "LiDAR", hint: "point cloud explorer" },
      { href: "/map", label: "HD map", hint: "fused map and provenance" },
      { href: "/inertial", label: "Inertial", hint: "ego-state timeline and events" },
    ],
  },
  {
    id: "verdyx", label: "VERDYX", role: "slice evaluation: per-slice regression + champion / challenger verdict",
    glyph: "VDX", order: 5, gate: "eval", flywheelStage: 7, home: "/training",
    nav: [
      { href: "/training", label: "Training", hint: "training jobs and model runs" },
      { href: "/govern", label: "Govern", hint: "loop control and championship" },
      { href: "/quality", label: "Quality", hint: "gold sets and slice metrics" },
      { href: "/analytics", label: "Analytics", hint: "corpus health and loop signal" },
    ],
  },
  {
    id: "forgyx", label: "FORGYX", role: "edge optimization: quantize, compile, benchmark, gate on latency and accuracy",
    glyph: "FGX", order: 6, gate: "benchmark", flywheelStage: 8, home: "/datasets",
    nav: [
      { href: "/datasets", label: "Datasets", hint: "sealed dataset delivery" },
      { href: "/forgyx", label: "Deployments", hint: "benchmark matrix and artifacts (M8)" },
    ],
  },
];

export const PLATFORMS_ORDERED = [...PLATFORMS].sort((a, b) => a.order - b.order);

export function platformById(id: string): Platform | undefined {
  return PLATFORMS.find((p) => p.id === id);
}

export function platformForPath(path: string): Platform | undefined {
  // Longest home/nav prefix wins, so /training resolves to VERDYX rather than the Labelox root.
  let best: Platform | undefined;
  let bestLen = -1;
  for (const p of PLATFORMS) {
    for (const href of [p.home, ...p.nav.map((n) => n.href)]) {
      if (href !== "/" && path.startsWith(href) && href.length > bestLen) {
        best = p;
        bestLen = href.length;
      }
    }
  }
  return best ?? platformById("labelox");
}
